chirp/drivers/anytone_clone.py

```diff
@@ -0,0 +1,269 @@
+# Copyright 2023 Dan Smith <dsmith@danplanet.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from builtins import bytes
+import struct
+import logging
+from abc import ABC, abstractmethod
+
+from chirp import util, errors
+
+LOG = logging.getLogger(__name__)
+
+
+# ############################################################################
+#                                                 Radio Communication Protocol
+# ############################################################################
+
+# TODO Consider lifting this to common for all drivers to utilize
+class RadioCommunicationProtocol(ABC):
+    """Abstract base class for encapsulating the logic that communicates with
+    the radio -- while remaining agnositic as to what content is being
+    communicated.
+
+    Usage example (within the download code, similar for upload):
+        with MyProtocol(pipe, block_size, file_ident) as protocol:
+            # `protocol.start_session()` is automatically called.
+            # `protocol.verify_radio_ident()` is automatically called.
+
+            for ...
+                for ...
+                    block = protocol.read_block(addr)
+                    data += block
+
+            # `protocol.end_session()` is automatically called.
+
+    """
+
+    def __init__(
+        self, pipe, block_size: int, file_ident, echos_write=True
+    ) -> None:
```

Comment:
I think I've mentioned this on the list, but surely many times in other PRs. I absolutely detest black formatting style (as do many others). I'm sorry you've done this work in this way and will have to reformat it, but that's how it is. I have to deal with it at work, but I get to choose here. These parens on their own lines are one of the ugliest aspects of it, IMHO.

Most of the rest of the chirp code is fairly conventional python formatting. Please reformat this to look like the rest of the codebase.

Reply:
If not Black, then which tool should I use? Any special settings?

Reply:
I can't offer you a tool to automatically format the code in more conventional python and chirp style, as you know. You can start by writing code that the current checker (flake8) will allow, and not adding arbitrary vertical stacking of things that will fit perfectly fine on fewer lines. Anywhere there's a closing paren on a line by itself is an easy regex for not-okay-ness :)